source/json/implementation/vss-json-implementation-big_integers.adb

```diff
@@ -0,0 +1,599 @@
+         Aux_L := Self.Data (Self.Last);
+         Aux_P := Self.Data (Self.Last - 1);
+
+         L := clz (Aux_L);
+
+         if L = 0 then
+            Truncated := Aux_P /= 0;
+            Value := Aux_L;
+
+         else
+            R := 64 - L;
+            Truncated := Interfaces.Shift_Left (Aux_P, Natural (L)) /= 0;
+            Value :=
+              Interfaces.Shift_Left (Aux_L, Natural (L))
+              or Interfaces.Shift_Right (Aux_P, Natural (R));
+         end if;
+
+         Truncated :=
+           @ or VSS.JSON.Implementation.Big_Integers.Non_Zero (Self, 2);
+      end if;
+   end Get_High_64;
+
+   --------------
+   -- Multiply --
+   --------------
+
+   procedure Multiply
+     (Self  : in out Big_Integer;
+      Value : Interfaces.Unsigned_64)
+   is
+      Carry : Interfaces.Unsigned_64 := 0;
+
+   begin
+      for J in Self.Data'First .. Self.Last loop
+         Scalar_Multiply (Self.Data (J), Value, Self.Data (J), Carry);
+      end loop;
+
+      if Carry /= 0 then
+         Push (Self, Carry);
+      end if;
+   end Multiply;
+
+   procedure Multiply
+     (Self : in out Big_Integer;
+      Y    : Limb_Array);
+
+   --------------
+   -- Multiply --
+   --------------
+
+   procedure Multiply
+     (Self : in out Big_Integer;
+      Y    : Limb_Array) is
+   begin
+      if Y'Length = 1 then
+         Multiply (Self, Y (Y'First));
+
+      else
+         declare
+            ZS : constant Big_Integer := Self;
+            Y0 : Limb_Type;
+            YJ : Limb_Type;
+            ZI : Big_Integer;
+
+         begin
+            if Y'Length /= 0 then
+               Y0 := Y (Y'First);
+               Multiply (Self, Y0);
+
+               for J in Y'First + 1 .. Y'Last loop
+                  YJ := Y (J);
+
+                  if YJ /= 0 then
+                     Set (ZI, ZS.Data (ZS.Data'First .. ZS.Last));
+                     Multiply (ZI, YJ);
+                     Add (Self, ZI.Data (ZI.Data'First .. ZI.Last), J);
+                  end if;
+               end loop;
+            end if;
+
+            Normalize (Self);
+         end;
+      end if;
+   end Multiply;
+
+   ----------------------
+   -- Multiply_Power_2 --
+   ----------------------
+
+   procedure Multiply_Power_2
+     (Self : in out Big_Integer; Exponent : Interfaces.Integer_32) is
+   begin
+      Shift_Left (Self, Exponent);
+   end Multiply_Power_2;
+
+   ----------------------
+   -- Multiply_Power_5 --
+   ----------------------
+
+   procedure Multiply_Power_5
+     (Self : in out Big_Integer; Exponent : Interfaces.Integer_32)
+   is
+      Large_Step       : constant := 135;
+      Large_Power_Of_5 : constant Limb_Array (0 .. 4) :=
+        (1414648277510068013, 9180637584431281687, 4539964771860779200,
+         10482974169319127550, 198276706040285095);
+
+      Small_Step       : constant := 27;
+      Small_Power_Of_5 : constant Limb_Type := 7450580596923828125;
+
+      Tail_Power_Of_5  : constant Limb_Array (0 .. Small_Step) :=
```

Comment:
Is it possible to use next aggregate here?
```ada
   Tail_Power_Of_5  : constant Limb_Array (0 .. 27) :=
     (for J in 0 .. 27 => 5 ** Natural (J));
```

I'm not sure if this will be a static however :disappointed:

Reply:
I prefer explicit values for tables.